sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceManager.java

```diff
@@ -0,0 +1,282 @@
+            LOGGER.info("Marking services as uninstalling: {} (out of {} services)",
+                    finishedServiceNames, services.size());
+
+            for (String name : finishedServiceNames) {
+                AbstractScheduler currentService = services.get(name);
+                if (currentService == null) {
+                    LOGGER.warn("Service '{}' does not exist, cannot trigger uninstall", name);
+                    continue;
+                }
+                if (currentService instanceof UninstallScheduler) {
+                    // Already uninstalling
+                    LOGGER.warn("Service '{}' is already uninstalling, leaving as-is", name);
+                    continue;
+                }
+
+                // Convert the DefaultScheduler to an UninstallScheduler. It will automatically flag itself with an
+                // uninstall bit in its state store and then proceed with the uninstall. When the uninstall has
+                // completed, it will return FINISHED to its next offers() call, at which point we will remove it. If
+                // the scheduler process is restarted before uninstall has completed, the caller should have added it
+                // back via putService(). When it's added back, it should be have already been converted to an
+                // UninstallScheduler. See SchedulerBuilder.
+                AbstractScheduler uninstallScheduler = ((DefaultScheduler) currentService).toUninstallScheduler();
+                if (isRegistered) {
+                    // Don't forget to also initialize the new scheduler if relevant...
+                    uninstallScheduler.registered(false);
+                }
+                services.put(name, uninstallScheduler);
+            }
+        } finally {
+            rwlock.unlock();
+        }
+    }
+
+    /**
+     * Removes the specified services after they have completed uninstall. Any unknown service names are ignored.
+     *
+     * @return the number of services which are still present after this removal
+     */
+    public int removeServices(Collection<String> uninstalledServiceNames) {
+        rwlock.lock();
+        try {
+            LOGGER.info("Removing {} uninstalled service{}: {} (from {} total services)",
+                    uninstalledServiceNames.size(),
+                    uninstalledServiceNames.size() == 1 ? "" : "s",
+                    uninstalledServiceNames,
+                    services.size());
+
+            for (String serviceName : uninstalledServiceNames) {
+                services.remove(serviceName);
+                sanitizedServiceNames.remove(CommonIdUtils.toSanitizedServiceName(serviceName));
+            }
+
+            return services.size();
+        } finally {
+            rwlock.unlock();
+        }
+    }
+
+    /**
+     * Gets a shared/read lock on the underlying data store, and returns all available services.
+     * Upstream MUST call {@link #unlockServices()} after finishing.
```

Comment:
I've noted elsewhere that I'm a little concerned about how long this lock can be held.  We should see if we can do something about that.